### PR TITLE
Issue-12704: added new necessary indexes.

### PR DESCRIPTION
**Pull request by LegateJD**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes 

### Description

Added absent indexes that were mentioned in [#12704](https://github.com/umbraco/Umbraco-CMS/issues/12704) issue.

[IX_umbracoContentVersion_NodeIdV2] index for [dbo].[umbracoContentVersion] table
image

[IX_umbracoContentVersionCultureVariation] index for [dbo].[umbracoContentVersionCultureVariation] table
image

[IX_umbracoNode_ObjectType_trashed_sorted] index for [dbo].[umbracoNode] table
image

[IX_umbracoRedirectUrl_culture_hash] index for [dbo].[umbracoRedirectUrl] table
image

Added possibility to assign multiple `IndexAttributes` to DTO properties.
image

### Steps to reproduce
1. Set up Umbraco CMS.
2. Check indexes via SSMS or with query 
`SELECT 
     TableName = t.name,
     IndexName = ind.name,
     IndexId = ind.index_id,
     ColumnId = ic.index_column_id,
     ColumnName = col.name,
     ind.*,
     ic.*,
     col.* 
FROM 
     sys.indexes ind 
INNER JOIN 
     sys.index_columns ic ON  ind.object_id = ic.object_id and ind.index_id = ic.index_id 
INNER JOIN 
     sys.columns col ON ic.object_id = col.object_id and ic.column_id = col.column_id 
INNER JOIN 
     sys.tables t ON ind.object_id = t.object_id 
WHERE 
     ind.is_primary_key = 0 
     AND ind.is_unique = 0 
     AND ind.is_unique_constraint = 0 
     AND t.is_ms_shipped = 0 
ORDER BY 
     t.name, ind.name, ind.index_id, ic.is_included_column, ic.key_ordinal;`
by listing all indexes.

### CODE REVIEW

File: src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionDto.cs [hunk 1]
```diff
@@ -20,6 +20,7 @@ public class ContentVersionDto
     [Column("nodeId")]
     [ForeignKey(typeof(ContentDto))]
     [Index(IndexTypes.NonClustered, Name = "IX_" + TableName + "_NodeId", ForColumns = "nodeId,current", IncludeColumns = "id,versionDate,text,userId")]
+    [Index(IndexTypes.NonClustered, Name = "IX_" + TableName + "_NodeIdV2", ForColumns = "nodeId,current", IncludeColumns = "versionDate,preventCleanup,text,userId")]
     public int NodeId { get; set; }
```
Comment (nzdev) [inline src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionDto.cs:24]: This one just needs the prevent cleanup column added. No need to have both indexes

File: src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionCultureVariationDto.cs [hunk 2]
```diff
@@ -19,6 +19,7 @@ internal class ContentVersionCultureVariationDto
     [Column("versionId")]
     [ForeignKey(typeof(ContentVersionDto))]
     [Index(IndexTypes.UniqueNonClustered, Name = "IX_" + TableName + "_VersionId", ForColumns = "versionId,languageId")]
+    [Index(IndexTypes.NonClustered, Name = "IX_" + TableName, ForColumns = "versionId", IncludeColumns = "id,languageId,name,date,availableUserId")]
```
Comment (nzdev) [inline src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionCultureVariationDto.cs:22]: The existing index could be used here if the columns are included

File: src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionDto.cs (hunk 1 above)
Comment (LegateJD) [reply src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionDto.cs:24]: @nzdev , then which name should be assigned to this index: IX_umbracoContentVersion_NodeIdV2 or just IX_umbracoContentVersion_NodeId?

File: src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionCultureVariationDto.cs (hunk 2 above)
Comment (LegateJD) [reply src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionCultureVariationDto.cs:22]: @nzdev , same question as for previous comment: which name should be used for index?

File: src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionDto.cs (hunk 1 above)
Comment (nzdev) [reply src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionDto.cs:24]: IX_umbracoContentVersion_NodeId.  Could delete the existing index and recreate it.

File: src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionCultureVariationDto.cs (hunk 2 above)
Comment (nzdev) [reply src/Umbraco.Infrastructure/Persistence/Dtos/ContentVersionCultureVariationDto.cs:22]: Original.

**Comment by LegateJD** [comment]
@nzdev , please, review.

File: src/Umbraco.Infrastructure/Migrations/Upgrade/V_10_2_0/TablesIndexesImprovement.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+public class TablesIndexesImprovement : MigrationBase
+{
+    public TablesIndexesImprovement(IMigrationContext context) : base(context)
+    {
+    }
+
+    protected override void Migrate()
+    {
+        AlterColumn<PropertyDataDto>(Constants.DatabaseSchema.Tables.PropertyData, "textValue");
```
Comment (nzdev) [inline src/Umbraco.Infrastructure/Migrations/Upgrade/V_10_2_0/TablesIndexesImprovement.cs:14]: Might be worth moving the alter columns to a separate pr to split from indexes work. It will need to be tested with existing dbs.

**Comment by nzdev** [comment]
Looks good to me.

**Comment by LegateJD** [comment]
@nzdev , can you approve this PR by clicking check mark for this PR?

**Review by nzdev — DISMISSED** [review]

File: src/Umbraco.Infrastructure/Persistence/Dtos/PropertyDataDto.cs [hunk 4]
```diff
@@ -23,7 +23,7 @@ internal class PropertyDataDto
 
     [Column("versionId")]
     [ForeignKey(typeof(ContentVersionDto))]
-    [Index(IndexTypes.UniqueNonClustered, Name = "IX_" + TableName + "_VersionId", ForColumns = "versionId,propertyTypeId,languageId,segment")]
+    [Index(IndexTypes.NonClustered, Name = "IX_" + TableName + "_VersionId", ForColumns = "versionId,propertyTypeId,languageId,segment", IncludeColumns = "varcharValue")]
```
Comment (nzdev) [inline src/Umbraco.Infrastructure/Persistence/Dtos/PropertyDataDto.cs:26]: This would need to remain a unique index.

Comment (ronaldbarendse) [reply src/Umbraco.Infrastructure/Persistence/Dtos/PropertyDataDto.cs:26]: Adding the `varcharValue` will significantly increase the index size and time to insert/update property data: why is this needed?

Comment (ronaldbarendse) [reply src/Umbraco.Infrastructure/Persistence/Dtos/PropertyDataDto.cs:26]: TIL: I see it's added as included column, so it's not included in the index key columns (see [Create indexes with included columns](https://learn.microsoft.com/en-us/sql/relational-databases/indexes/create-indexes-with-included-columns?view=sql-server-ver16)).

Comment (nzdev) [reply src/Umbraco.Infrastructure/Persistence/Dtos/PropertyDataDto.cs:26]: Perhaps revert this one line for now @ronaldbarendse ?

File: src/Umbraco.Infrastructure/Persistence/Dtos/NodeDto.cs [hunk 5]
```diff
@@ -27,6 +27,7 @@ public class NodeDto
     [Column("parentId")]
     [ForeignKey(typeof(NodeDto))]
     [Index(IndexTypes.NonClustered, Name = "IX_" + TableName + "_ParentId")]
+    [Index(IndexTypes.NonClustered, Name = "IX_" + TableName + "_parentId_nodeObjectType", ForColumns = "parentID,nodeObjectType", IncludeColumns = "trashed,nodeUser,level,path,sortOrder,uniqueID,text,createDate")]
```
Comment (nzdev) [inline src/Umbraco.Infrastructure/Persistence/Dtos/NodeDto.cs:30]: _parentId_nodeObjectType covers _ParentId. replace / only _parentId_nodeObjectType needed.

File: src/Umbraco.Infrastructure/Persistence/Dtos/PropertyDataDto.cs (hunk 4 above)
Comment (JasonElkin) [reply src/Umbraco.Infrastructure/Persistence/Dtos/PropertyDataDto.cs:26]: Just trying to resurrect this PR. @ronaldbarendse @nzdev, can you confirm if this line should be reverted?

File: src/Umbraco.Infrastructure/Migrations/Upgrade/V_10_2_0/TablesIndexesImprovement.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
+﻿using Umbraco.Cms.Core;
+using Umbraco.Cms.Infrastructure.Persistence.Dtos;
+
+namespace Umbraco.Cms.Infrastructure.Migrations.Upgrade.V_10_2_0;
```
Comment (JasonElkin) [inline src/Umbraco.Infrastructure/Migrations/Upgrade/V_10_2_0/TablesIndexesImprovement.cs:4]: This will now need to target v_12_1_0

File: src/Umbraco.Infrastructure/Migrations/Upgrade/UmbracoPlan.cs [hunk 7]
```diff
@@ -291,6 +291,7 @@ protected void DefinePlan()
         To<AddMemberPropertiesAsColumns>("{B7E0D53C-2B0E-418B-AB07-2DDE486E225F}");
 
         // TO 10.2.0
+        To<TablesIndexesImprovement>("{1187192D-EDB5-4619-955D-91D48D738871}");
```
Comment (JasonElkin) [inline src/Umbraco.Infrastructure/Migrations/Upgrade/UmbracoPlan.cs:294]: This will now need moving to below `// TO 12.1.0`

**Review by JasonElkin — CHANGES_REQUESTED** [review]

**Comment by JasonElkin** [comment]
Hey @LegateJD, thanks for contributing this! 
It looks like this PR gotten stuck - let's see if we can unstick it and get this merged in!

While we wait for feedback on [this comment](https://github.com/umbraco/Umbraco-CMS/pull/12756#discussion_r1196263642), would you be able to make the rest of the necessary changes:

1. Remove the _ParentId index [as suggested by nzdev](https://github.com/umbraco/Umbraco-CMS/pull/12756#discussion_r974363339)
2. Fix the conflicts (sorry)
3. Update references from 10.2.0 to 12.1.0, which this PR will target if we manage to get this merged soon.

Thanks,
Jason

**Comment by JasonElkin** [comment]
Super, thanks @LegateJD!

I'll get this tested and get back to you as soon as I can.

**Review by JasonElkin — APPROVED** [review]

**Comment by JasonElkin** [comment]
Tested and merged! 🚀🎉

Thanks again @LegateJD and @nzdev for this - I'll go ahead and close the issue too.
